### PR TITLE
Add `second_factors`

### PR DESCRIPTION
**Pull request by Joerger**
Add `second_factors` and prefer it over `second_factor`.

We don't currently plan on removing second_factor, as this would require a more complicated migration process. Instead we will just derive `second_factors` from `second_factor` (and vice versa) and output a warning log when `second_factor` is set.

There is no plan to deprecate `second_factor` completely. When `second_factor` is set and `second_factors` is not, or vice versa, we convert from one to the other.

In a follow up PR I will update as much logic as possible to use `second_factors` instead of `second_factor`, as they are two sources of the same information.

In this PR I've also added the SSO second_factor type. It is currently completely unused, but we'd rather get the proto changes into v17 rather than waiting until SSO MFA is fully released in a minor version.

Follow up TODO: Update docs.

Changelog: Add new second_factors field to cluster auth preference for more clarity and granularity over which 2fa methods are enabled in a cluster.

Depends on https://github.com/gravitational/teleport/pull/47230

### CODE REVIEW

File: api/types/authentication.go
Comment (Joerger) [inline api/types/authentication.go:87]: note: this will be used in the follow up PR, but I'd rather add it here with these changes.

File: api/proto/teleport/legacy/types/types.proto
Comment (tigrato) [inline api/proto/teleport/legacy/types/types.proto:2104]:
```suggestion
  repeated SecondFactorType allowed_second_factors = 21 [(gogoproto.jsontag) = "supported_second_factors,omitempty"];
```

File: api/types/authentication.go
Comment (tigrato) [inline api/types/authentication.go:355]: if our goal is to migrate away from string, should we convert the second factor to SecondFactors?

File: api/types/authentication.go
Comment (tigrato) [inline api/types/authentication.go:393]: since SecondFactor takes precedence, should we also handle it here?

File: api/proto/teleport/legacy/types/types.proto
Comment (Joerger) [reply api/proto/teleport/legacy/types/types.proto:2104]: @rosstimothy WDYT about `supported_second_factors`? Product already approved `second_factors` right?

File: lib/config/fileconf.go
Comment (codingllama) [inline lib/config/fileconf.go:1109]: Similar to other comments, won't this cause changes to CAP?

**Review by codingllama — COMMENTED** [review]
Thanks for the separate PR, Brian! Apologies for the delay.

File: api/types/authentication.go
Comment (codingllama) [reply api/types/authentication.go:748]: I think the comment still applies for second_factor="otp".

File: lib/config/fileconf.go
Comment (codingllama) [reply lib/config/fileconf.go:1097]: even so I'm of the option it should stay, but I'll yield to UX as usual.

File: api/types/authentication.go
Comment (Joerger) [reply api/types/authentication.go:398]: No, we're just going to prefer WebAuthn > SSO > TOTP rather than an ordered list.

File: lib/config/fileconf.go
Comment (Joerger) [reply lib/config/fileconf.go:1109]: I'm not sure what you mean

Comment (codingllama) [reply lib/config/fileconf.go:1109]: I mean that we are writing both second_factor and second_factors to the "cluster_auth_preference". This means that, on a read, we'll return a resource that has both fields set and thus is invalid by Teleport's own criteria.

This is all to say I would rather keep the cap as the user wrote it and calculate whatever we need in the getters, on the fly.

File: api/types/authentication.go
Comment (Joerger) [reply api/types/authentication.go:382]: I think we should just validate `SecondFactor` in `CheckAndSetDefaults`.

File: api/types/authentication.go
Comment (codingllama) [inline api/types/authentication.go:336]: Similar to other comments, should we set only the SecondFactor (singular) field here?

File: api/types/second_factor.go
Comment (codingllama) [inline api/types/second_factor.go:115]: Log the error?

File: api/types/second_factor.go
Comment (codingllama) [inline api/types/second_factor.go:130]:
```suggestion
		return "", trace.BadParameter("invalid SecondFactorType value: %q", *s)
```

So we avoid capitalization at the start. Same for others.

- https://go.dev/wiki/CodeReviewComments#error-strings

File: api/types/second_factor.go
Comment (codingllama) [inline api/types/second_factor.go:160]:
```suggestion
		return trace.BadParameter("invalid RequireMFAType type: %T", val)
```

File: lib/config/fileconf.go
Comment (Joerger) [reply lib/config/fileconf.go:1097]: This _is_ taking the cap as the user wrote it. But I will remove the warning log and update `CheckAndSetDefaults` to return `trace.BadParameter("must set either SecondFactor or SecondFactors, not both")` sot that this would return an error instead.

File: api/types/second_factor.go
Comment (codingllama) [inline api/types/second_factor.go:168]:
```suggestion
		return trace.BadParameter("invalid SecondFactorType enum: %v", val)
```

File: api/types/second_factor.go
Comment (codingllama) [inline api/types/second_factor.go:166]: Use `s` for the receiver, like the other methods?

File: lib/config/fileconf.go
Comment (codingllama) [reply lib/config/fileconf.go:1097]: Right, sorry for the noise.

File: api/types/authentication.go
Comment (codingllama) [reply api/types/authentication.go]: Would erroring here make for a clearer error?

Comment (Joerger) [reply api/types/authentication.go]: I don't think so

Comment (Joerger) [reply api/types/authentication.go]: bf78d9e0232e748f6b323ac7895dbba85f3040ff

**Review by codingllama — APPROVED** [review]

File: api/types/authentication.go
Comment (codingllama) [inline api/types/authentication.go:90]:
```suggestion
	// IsSecondFactorTOTPAllowed checks if users can use TOTP as an MFA method.
```

File: api/types/authentication.go
Comment (codingllama) [inline api/types/authentication.go:778]: Could we get test coverage for this?

File: api/types/authentication.go
Comment (codingllama) [inline api/types/authentication.go:729]: The `len` check is usually better, `[]SecondFactorType{}` is not nil but it has zero length.

File: api/types/second_factor.go
Comment (codingllama) [inline api/types/second_factor.go:35]: This is an interesting semantical change. Up to this point SecondFactorOn meant that all configs had to be set, but now we could have only SSO second factor but have to call it "on" (because we lack a value to express it).

I think this is (mostly?) fine (we typically check for "otp available" or "webauthn config exists" as well), but we should update godocs on SecondFactorOn to mention the new corner-case.

We should also test this scenario in practice when we get to it. Maybe leave a TODO here as well?

File: api/types/authentication.go
Comment (codingllama) [inline api/types/authentication.go:325]: Note that `!= nil` doesn't mean `len > 0`. A len check would be safer.

File: api/types/authentication.go
Comment (Joerger) [reply api/types/authentication.go:729]: Oops, I really thought this wasn't the case. I guess only `var sfs []SecondFactorType` would evaluate to nil.

File: lib/config/configuration_test.go
Comment (Joerger) [reply lib/config/configuration_test.go:1]: Is there any way for me to avoid this happening? Without gofumpt I get lint issues, but with it it makes unnecessary changes like this.

File: api/types/second_factor.go
Comment (Joerger) [reply api/types/second_factor.go:35]: The only purpose of this is here - https://github.com/gravitational/teleport/pull/47426/commits/82eff9a725ead23ce6ec50cd67ff6d471499da22

If you'd like I can hold off on merging this PR until that one is also ready and approved

Comment (Joerger) [reply api/types/second_factor.go:35]: I wanted to set this to "on" since we are relying on the logic in some places. However this will be unnecessary as soon as https://github.com/gravitational/teleport/pull/47426/commits/82eff9a725ead23ce6ec50cd67ff6d471499da22 is merged, so I'd rather not overcomplicate things. If you'd like I can hold off on merging this until that PR is ready and approved.

File: api/types/second_factor.go
Comment (Joerger) [reply api/types/second_factor.go:1]: Yes it should be covered well enough, especially once https://github.com/gravitational/teleport/pull/47426 to make `second_factors` used primarily.

File: api/types/authentication.go
Comment (Joerger) [reply api/types/authentication.go]: 6ab4fdc5560b7900cb21ab497d5970d3d9fa19c9

File: api/types/second_factor.go
Comment (codingllama) [reply api/types/second_factor.go]: I'm happy knowing that there is a follow up. I'd rather get this merged than wait.

File: api/types/authentication.go
Comment (codingllama) [inline api/types/authentication.go:781]: Why does `hasSSO` matter here? It looks like the test doesn't really care whether it's true of false.

File: api/types/authentication_test.go
Comment (codingllama) [inline api/types/authentication_test.go:94]: 1. Assert GetSecondFactors / GetSecondFactor directly? I think we care more about the flag than the derived settings. Same for others.

2. nit: Use `assert` instead of `require`?

File: api/types/authentication_test.go
Comment (codingllama) [inline api/types/authentication_test.go:136]: No asserts?

File: api/types/authentication_test.go
Comment (codingllama) [inline api/types/authentication_test.go:147]: That's a bit strange, I would expected we needed to pass `**trace.BadParameterError` to errors.As, since it's `*trace.BadParameter` that implements `error`.

File: api/types/authentication_test.go
Comment (codingllama) [inline api/types/authentication_test.go:72]:
```suggestion
func TestNewAuthPreference_secondFactors(t *testing.T) {
```

File: api/types/second_factor.go
Comment (codingllama) [inline api/types/second_factor.go:34]: nit:

```suggestion
	case hasOTP && hasWebAuthn:
```

File: api/types/second_factor_test.go
Comment (codingllama) [inline api/types/second_factor_test.go:25]:
```suggestion
```

This one is pretty clear already, imo.

File: api/constants/constants.go
Comment (codingllama) [inline api/constants/constants.go:212]: Is this still true given recent changes?

File: api/types/authentication_test.go
Comment (Joerger) [reply api/types/authentication_test.go:94]: I don't really care about `SecondFactor` here and `GetSecondFactors` just returns the list we set above. My intention with these tests was to test these default settings like passwordless. I'll add a separate test for converting SecondFactor to SecondFactors.

File: api/types/authentication.go
Comment (Joerger) [reply api/types/authentication.go:781]: The test does set `SecondFactors=[sso]`, but what we really care about is `if second factor is enforced && local second factors are disabled`. I'll add a `IsLocalMFAEnabled` helper to clarify things.

**Review by codingllama — APPROVED** [review]

File: lib/config/configuration_test.go
Comment (Joerger) [reply lib/config/configuration_test.go:1]: I just needed to add this to my vscode settings:

```
"[go]": {
    "editor.formatOnSaveMode": "modifications",
},
```

**Review by rosstimothy — APPROVED** [review]

**Comment by Joerger** [comment]
> @Joerger did we ever got docs updates for this one?

No not yet, it's on my TODO list, along with SSO MFA docs

**Pull request merged**
